sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala

```diff
@@ -87,8 +87,15 @@ case class CustomShuffleReaderExec private(
     Iterator(desc)
   }
 
-  def hasCoalescedPartition: Boolean =
-    partitionSpecs.exists(_.isInstanceOf[CoalescedPartitionSpec])
+  /**
+   * Returns true iff some non-empty partitions were combined
+   */
+  def hasCoalescedPartition: Boolean = {
+    partitionSpecs.exists {
+      case s: CoalescedPartitionSpec => s.endReducerIndex - s.startReducerIndex > 1
```

Comment:
I prefer use `s.startReducerIndex + 1 != s.endReducerIndex` since we have a special case `CoalescedPartitionSpec (0, 0, 0)`

Reply:
The intent of this PR is to have`hasCoalesedPartitions` only return true if some were combined - to avoid reporting "coalesced" when partition spec looks something like `CoalescePartitionSpec(0, 1), CoalescePartitionSpec(1, 2)`
Even if `CoalescePartitionSpec(0, 0, 0)` is possible, it's not combining anything.

Reply:
We coalesce all partitions to an empty partition if the input RDD is empty
```
// If all input RDDs have 0 partition, we create an empty partition for every shuffle reader.
if (validMetrics.isEmpty) {
  return Seq.fill(numShuffles)(Seq(CoalescedPartitionSpec(0, 0, 0)))
}
```

Reply:
If you  started with no partitions, you are not coalescing, i.e. you are not combining any partitions.

Reply:
Start with empty RDD does not mean the partition after shuffle is empty, instead the shuffle partitions is `spark.sql.shuffle.partitions` without AQE. After `CoalesceShufflePartitions`, the shuffle partitions are coalesced from `spark.sql.shuffle.partitions` to one empty partition. So it shouble be coalesced.

Reply:
Interesting.  I didn't realize Spark could produce `spark.sql.shuffle.partitions` empty partitions.

Reply:
```
// If all input RDDs have 0 partition, we create an empty partition for every shuffle reader.
if (validMetrics.isEmpty) {
  return Seq.fill(numShuffles)(Seq(CoalescedPartitionSpec(0, 0, 0)))
}
```

Shall we use `CoalescedPartitionSpec(0, numReducers, 0)`?

Reply:
Although `CoalescedPartitionSpec(0, 0,0 )` looks a little hack but it's safe and effective.

We don't know the numReducers if the shuffle comes from `EmptyRDD`, then we can only assume the number is same with `spark.sql.shuffle.partitions`. But sometimes the assuming is not correct, e.g. if serveral `EmptyRDD` has different pre-shuffle partition numbers.

Reply:
I see, then I agree `s.startReducerIndex + 1 != s.endReducerIndex` is better.

Reply:
@ulysses-you can you open a followup PR? This is really a corner case and I'd like to merge this PR first.

Reply:
@cloud-fan created https://github.com/apache/spark/pull/33342